src/main/java/com/godaddy/sonar/ruby/simplecovrcov/SimpleCovRcovJsonParserImpl.java

```diff
@@ -11,49 +16,78 @@
+        }
+        return coverageReport;
+    }
 
-            String filePath = ((Map.Entry) coverageJsonObj.entrySet().toArray()[j]).getKey().toString();
-            LOG.debug("filePath " + filePath);
+    private Map<String, CoverageMeasuresBuilder> processCoverageReport(CoverageReport coverageReport) {
+        if (coverageReport.hasSeveralReporters()) {
+            LOG.warn("Coverage report has several reporters! Notice, please, that only the first one will be considered by Sonar!");
+        }
+        return processReporter(coverageReport.getFirstReporter());
+    }
 
-            JsonArray coverageArray = coverageJsonObj.get(filePath).getAsJsonArray();
+    private Reporter buildReporter(Map.Entry coverageMapEntry) {
+        JsonObject coverageJsonObj = ((JsonObject)coverageMapEntry.getValue()).get("coverage").getAsJsonObject();
+        String reporterName = coverageMapEntry.getKey().toString();
+        Reporter reporter = new Reporter(reporterName);
+        for(Map.Entry reportItemMapEntry : coverageJsonObj.entrySet()) {
+            reporter.addItem(buildReporterItem(reportItemMapEntry));
+        }
+        return reporter;
+    }
 
-            // for each line in the coverage array
-            for (int i = 0; i < coverageArray.size(); i++) {
-                Long line = null;
+    private ReporterItem buildReporterItem(Map.Entry reporterItemMapEntry) {
+        String filename = reporterItemMapEntry.getKey().toString();
+        JsonArray marksJsonArr = ((JsonArray)reporterItemMapEntry.getValue());
+        Collection<Mark> marks = new ArrayList<>();
+        for(JsonElement marksEl : marksJsonArr) {
+            marks.add(new Mark(marksEl.toString(), marksEl.isJsonNull()));
+        }
+        return new ReporterItem(filename, marks);
+    }
 
-                if (!coverageArray.get(i).isJsonNull()) {
-                    line = coverageArray.get(i).getAsLong();
-                }
+    private Map<String, CoverageMeasuresBuilder> processReporter(Reporter reporter) {
+        Map<String, CoverageMeasuresBuilder> coveredFiles = Maps.newHashMap();
+        for (ReporterItem reporterItem : reporter.getItems()) {
+            CoverageMeasuresBuilder fileCoverage = processReporterItem(reporterItem);
+            coveredFiles.put(reporterItem.getFilename(), fileCoverage);
+        }
+        return coveredFiles;
+    }
 
-                Integer intLine = 0;
-                int lineNumber = i + 1;
-                if (line != null) {
-                    intLine = line.intValue();
-                    fileCoverage.setHits(lineNumber, intLine);
-                }
+    private CoverageMeasuresBuilder processReporterItem(ReporterItem reporterItem) {
+        CoverageMeasuresBuilder fileCoverage = CoverageMeasuresBuilder.create();
+        for (int markId = 0; markId < reporterItem.getMarks().size(); markId++) {
+            Mark mark = (Mark)reporterItem.getMarks().toArray()[markId];
```

Comment:
`Mark mark = (Mark)reporterItem.getMarks().toArray()` is not dependant on `makrId` and should be refactored outside the for-loop